packages/eslint-plugin/src/rules/no-redundant-type-constituents.ts

```diff
@@ -0,0 +1,323 @@
+  [ts.TypeFlags.TemplateLiteral]: ts.TypeFlags.String,
+} as const;
+
+const literalTypeFlags = [
+  ts.TypeFlags.BigIntLiteral,
+  ts.TypeFlags.BooleanLiteral,
+  ts.TypeFlags.NumberLiteral,
+  ts.TypeFlags.StringLiteral,
+  ts.TypeFlags.TemplateLiteral,
+] as const;
+
+const primitiveTypeFlags = [
+  ts.TypeFlags.BigInt,
+  ts.TypeFlags.Boolean,
+  ts.TypeFlags.Number,
+  ts.TypeFlags.String,
+] as const;
+
+const primitiveTypeFlagNames = {
+  [ts.TypeFlags.BigInt]: 'bigint',
+  [ts.TypeFlags.Boolean]: 'boolean',
+  [ts.TypeFlags.Number]: 'number',
+  [ts.TypeFlags.String]: 'string',
+} as const;
+
+type PrimitiveTypeFlag = typeof primitiveTypeFlags[number];
+
+interface TypeNodeWithValue {
+  literalValue: unknown;
+  typeNode: TSESTree.TypeNode;
+}
+
+function addToMapGroup<Key, Value>(
+  map: Map<Key, Value[]>,
+  key: Key,
+  value: Value,
+): void {
+  const existing = map.get(key);
+
+  if (existing) {
+    existing.push(value);
+  } else {
+    map.set(key, [value]);
+  }
+}
+
+function describeLiteralType(type: ts.Type): unknown {
+  return type.isStringLiteral()
+    ? JSON.stringify(type.value)
+    : type.isLiteral()
+    ? type.value
+    : util.isTypeTemplateLiteralType(type)
+    ? 'template literal type'
+    : util.isTypeBigIntLiteralType(type)
+    ? `${type.value.negative ? '-' : ''}${type.value.base10Value}n`
+    : tsutils.isBooleanLiteralType(type, true)
+    ? 'true'
+    : tsutils.isBooleanLiteralType(type, false)
+    ? 'false'
+    : 'literal type';
+}
+
+function isNodeInsideReturnType(node: TSESTree.TSUnionType): boolean {
+  return !!(
+    node.parent?.type === AST_NODE_TYPES.TSTypeAnnotation &&
+    node.parent.parent &&
+    util.isFunctionType(node.parent.parent)
+  );
+}
+
+/**
+ * @remarks TypeScript stores boolean types as the union false | true, always.
+ */
+function unionTypePartsUnlessBoolean(type: ts.Type): ts.Type[] {
+  return type.isUnion() &&
+    type.types.length === 2 &&
+    tsutils.isBooleanLiteralType(type.types[0], false) &&
+    tsutils.isBooleanLiteralType(type.types[1], true)
+    ? [type]
+    : tsutils.unionTypeParts(type);
+}
+
+export default util.createRule({
+  name: 'no-redundant-type-constituents',
+  meta: {
+    docs: {
+      description:
+        'Disallow members of unions and intersections that do nothing or override type information',
+      recommended: false,
+      requiresTypeChecking: true,
+    },
+    messages: {
+      literalOverridden: `{{literal}} is overridden by {{primitive}} in this union type.`,
+      primitiveOverridden: `{{primitive}} is overridden by the literal {{literal}} in this intersection type.`,
+      overridden: `'never' is overridden by other types in this {{container}} type.`,
+      overrides: `'{{typeName}}' overrides all other types in this {{container}} type.`,
+    },
+    schema: [],
+    type: 'suggestion',
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      TSIntersectionType(node): void {
```

Comment:
it'd be good if we could micro-optimise this a bit by handling the "ast determinable" cases without types. EG: `foo | any` and `bar | unknown` won't need type info to analyse.

There's also a lot of cases we could handle without type info like `string`, `number`, etc - so  it'd be good if we could limit this rule to only using type information if and only if it's a reference type.

Type info is expensive to access due to the lazy calculation - so if we can avoid calling `getTypeChecker` until we really need it - that'd be great. I can see a rule like this would be usper expensive otherwise.

----

I would also suggest that maybe you do this on `TSIntersectionType:exit` instead of on enter.
That way you can analyse the child intersections first (in cases like `(foo | bar) | baz`), mark them as analysed with result `any | not-any`, and then save double-checking them.

Reply:
> `TSIntersectionType:exit`

I want to say this should already be cached by the type checker, so the caching is already done for us? Am I missing the point @bradzacher?

Reply:
I don't think that the checker does quite as much caching as you might want or think it does. At least from my observation during debugging - the methods do a fair amount of recalculation when called.

With a healthy dose of caching and short-circuiting we should be able to squeeze out some better perf!

Reply:
Well that's unfortunate 😞, I was hoping we could get away without recreating structure on our own.

As a followup to these changes I'm thinking of filing an issue about doing a broader audit of places that could go with similar logic: checking keyword type of nodes to bypass using the type checker altogether.